src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -53,40 +53,61 @@ public Assembly LoadFromPath(string fullPath)
 
         private Assembly LoadFromPathUnchecked(string fullPath)
         {
+            var assemblyIdentity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
+            return LoadFromPathUncheckedCore(fullPath, assemblyIdentity);
+        }
+
+        private Assembly LoadFromPathUncheckedCore(string fullPath, AssemblyIdentity identity)
+        {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
 
+            // Check if we have already loaded an assembly with the same identity or from the given path.
+            Assembly loadedAssembly = null;
             lock (_guard)
             {
                 Assembly existingAssembly;
                 if (_loadedAssembliesByPath.TryGetValue(fullPath, out existingAssembly))
                 {
-                    return existingAssembly;
+                    loadedAssembly = existingAssembly;
+                }
+                else if (identity != null && _loadedAssembliesByIdentity.TryGetValue(identity, out existingAssembly))
+                {
+                    loadedAssembly = existingAssembly;
                 }
             }
```

Comment:
We need to ensure the same assembly is added to both path and identity cache, AddToCache below does that and returns loaded assembly.